### PR TITLE
Adapting `CutoutFactory` to account for error-less TICA Cubes

### PR DESCRIPTION
**Pull request by jaymedina**
### Summary

This PR updates `CutoutFactory` to account for the removed error array in `TicaCubeFactory` (work that was done in  #87). It involves turning the `product` kwarg from `cube_cut()` into a class attribute so that it doesn't have to be fed into every method as an argument before being called. It also involves using this new `self.product` class attribute to make conditionals that make the code dynamic and changes the logic depending on whether the product for the cutouts is `"TICA"` or `"SPOC"`. Unit test changes are also made where appropriate.

----

### Task List

- [x] Account for removed error array in `CutoutFactory`
- [x] Update `cube_cut` unit tests
- [x] Rebase onto #87

----

### Screenshots

### CODE REVIEW

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:683]: (this comment is for line number 685, GH doesn't allow comments on collapsed sections of code)

`uncert_cube` docstring needs to be updated

in general, we should update the docstrings everywhere we have changed the signature of the functions

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:851]: why is `product` an argument of cube_cut() function still? 

self.product should only be assigned during initialization of the cutout factory, now, right?

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:908]:
```suggestion
                img_cutout, uncert_cutout, aperture = results
```

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:910]:
```suggestion
                img_cutout, aperture = results
```

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:469]: returning different amounts of things instead of always the same amount of things makes logic of dealing with the return values much harder.

the logic of dealing with spoc vs. tica is spreading throughout the codebase.

what do you think about initializing `uncert_cutout` to `None` and return all three variables no matter what?

we also need to update the docstring about the return value with whatever this function ends up doing.

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:675]: if when we make the cutout, it returns `None` for `uncert_cube`, we can call this exactly the same for both TICA and SPOC and we can leave uncert_cube as a required param. then we don't need another conditional when we call this function, since for TICA `None` will be passed in.

File: astrocut/make_cube.py
Comment (falkben) [inline astrocut/make_cube.py:441]: this comment should be updated, maybe include something about the reason it is 1 instead of 2

**Review by falkben — CHANGES_REQUESTED** [review]
tesscut code will need to change for this to work

astroquery? anything else?

File: astrocut/cube_cut.py
Comment (jaymedina) [reply astrocut/cube_cut.py:849]: `self.product` is assigned to `None` when the `CutoutFactory` class is initialized. It's then replaced by "TICA" or "SPOC" depending on what value the user passes when calling `cube_cut`. But this reminds me, I should enforce that the "TICA" or "SPOC" string is in all uppercase so we don't run into a case-sensitive bug down the line.

**Comment by jaymedina** [comment]
> tesscut code will need to change for this to work
> 
> astroquery? anything else?

`astroquery.mast.Tesscut` wraps around the TESSCut API, so once we redeploy TESSCut API, we should be good to go (I will keep an eye on unit tests). We also have to ensure that [this notebook](https://github.com/spacetelescope/mast_notebooks/blob/main/notebooks/astrocut/making_tess_cubes_and_cutouts/making_tess_cubes_and_cutouts.ipynb) runs correctly, too. More detail is in the [Jira ticket](https://jira.stsci.edu/browse/ASB-22309) for this Epic.

File: astrocut/cube_cut.py
Comment (falkben) [reply astrocut/cube_cut.py]: i see, i missed that originally -- i guess i'd expected it to be an argument to the initialization of the coutoutfactory instance.

in general, i think this changeset is moving pretty far away from the idea of a factory class and we're now adding a bunch of conditionals if / else throughout the codebase which i'm not a huge fan of. 

perhaps we should hold off on refactoring the product and first just implement the change to remove the error array from TICA cubes?

When refactoring, i think there might be some ways to try and avoid all the conditional logic that is in this PR, which gets very hard to maintain.

Comment (jaymedina) [reply astrocut/cube_cut.py]: We've already removed the error array from the TICA cubes in PR #87 

I ran some tests and confirmed that the removal of the error arrays from the TICA cubes is worth the hassle of refactoring things in `CutoutFactory` (cubes are now 1/2 the size and take about 1/3 less time to make). I thought about making a new class just for TICA called `TicaCutoutFactory`, the same way things exist separately in `make_cube.py`, but it leads to a lot of redundancies.

Comment (falkben) [reply astrocut/cube_cut.py]: sorry, i mean cutouts, not cubes, above.

i'm not disagreeing about removing the error component from the cutout (or the cube). 

i wonder, do fits files support sparse arrays?

Comment (jaymedina) [reply astrocut/cube_cut.py]: You can make sparse matrices as numpy arrays, so I assume if `fits` can support a numpy array, they can support a sparse array if numpy is its data type.

Comment (falkben) [reply astrocut/cube_cut.py]: I count 11 conditionals where we're checking if the cutout product is TICA or SPOC when making a cutout and I'm worried about the added complexity that has been added to this codebase by adding just a single new product type.

What I would like is a lot less conditional behavior based on the product and to instead utilize the factory based structure this module purports to be to avoid all these conditionals. 

We can handle this in a separate PR / ticket if you like. I do think this is important for maintainability, both for the cube creation as well as the cutouts.  I did mention some of this when we added TICA to the cube creation module as well.

With some refactoring, we can possibly simplify this to use the factory methodology and define objects (SPOC / TICA) that use the factory to make the cutouts, and can keep the different behavior isolated to their own class methods.  If you like, we can work together to sketch this out.

I think we should not undertake a large refactoring in this PR, so I will instead focus my comments on the specifics of not returning the error array when making TICA cutouts.

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:405]: should this line change as well? the second item is either a numpy array or None.

some type annotations on these functions would also help a lot. i think it would be valuable to add mypy at some point to this project and to type annotate as much as we can.

File: astrocut/tests/test_cube_cut.py
Comment (falkben) [inline astrocut/tests/test_cube_cut.py:370]: can we check that FLUX_ERR is present for SPOC and not present in TICA?

**Review by falkben — CHANGES_REQUESTED** [review]

File: astrocut/cube_cut.py
Comment (jaymedina) [reply astrocut/cube_cut.py]: I see what you mean, and yes it makes sense to do this refactoring in a separate PR. This PR has a smaller time-constraint, as we have to start generating the errorless TICA cubes as soon as Aug. 14th, so I'd like to have this merged by the end of this week (Aug. 3rd) ideally. That way @hmlewis-astro can start generating the error-free TICA cubes asynchronously while I work on simplifying the code. Let me know if you agree with this and I'll make a new Jira ticket for the refactoring @falkben

Comment (falkben) [reply astrocut/cube_cut.py]: That sounds good. 

Also, this might be related: https://github.com/spacetelescope/astrocut/issues/62 since there was a similar discussion when the make cube stuff was added for TICA https://github.com/spacetelescope/astrocut/pull/59#discussion_r890526958

I'm not sure what makes sense re: refactoring, whether class inheritance would actually make things more maintainable etc, but we might want to have these work items linked, somehow.  

I'd be happy to sketch this out with you at some point to try and generalize this a bit.

Comment (jaymedina) [reply astrocut/cube_cut.py]: OK, I will link all three PRs (#88, #62, and #59) in the Astrocut Refactoring Jira ticket. Thank you!

**Comment by jaymedina** [comment]
> There's some flake8 failures in CI -- look relatively trivial to fix: https://github.com/spacetelescope/astrocut/actions/runs/5716882528/job/15489424621?pr=88

Interesting - I haven't seen that codestyle error before. I'll update with a new commit.

**Review by falkben — DISMISSED** [review]
regarding the [CI failure](https://github.com/spacetelescope/astrocut/actions/runs/5717424375/job/15491139878?pr=88) for python 3.11 and installing everything from source, I think we might need an additional step to install openblas: `libblas-dev`

this can be done in a different PR, though. Once this is merged, I don't mind taking a look.

**Review by falkben — APPROVED** [review]

**Comment by jaymedina** [comment]
Before we merge this into main: 

I demo'ed the work here earlier this week and got a suggestion from @scfleming to keep the `FLUX_ERR` cutouts in TICA for the sake of making the code easier to integrate into 3rd party software like `Lightkurve` and user scripts. I believe this would involve changing [the logic that starts here](https://github.com/spacetelescope/astrocut/blob/7d488c5842144054adae76d1b4099006d8b3a3d9/astrocut/cube_cut.py#L733) in this branch so that an empty array (`empty_arr`) gets filled in for each row of this column when the product is TICA, and also updating the unit tests accordingly (we will check TICA for the existence of `FLUX_ERR` again, and ensure that they are empty arrays). Let me know what you all think of this logic before I make a commit @falkben @hmlewis-astro

**Comment by falkben** [comment]
I'd be happy with that change @jaymedina -- it will make this cutout code easier to maintain, as well.

It will make the cutouts the same size as the SPOC cutouts, though, (not any smaller) but I don't think that was the intention with any of this.

**Comment by jaymedina** [comment]
> It will make the cutouts the same size as the SPOC cutouts, though, (not any smaller) but I don't think that was the intention with any of this.

Updated accordingly. Yep, the cutouts themselves are small, so it's not as big of a deal for them to have the error array. The important thing was to reduce the size of the cubes by removing the error array _there_, and to adapt the `CutoutFactory` code to take this change into account. Thanks for the quick response!

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:732]: i'm a little uncertain about this. numpy array assignment is done by reference. i'm just worried that because we are using empty_arry here to assign into cadence_arry, or appending to columns, that something might change inside of `empty_arr` and then would be applied to flux_err_arry at this point.

e.g. to illustrate what i'm worried about

```python
>>> import numpy as np
>>> empty_arr = np.zeros((2,3))
>>> other_arr = empty_arr[:,1]
>>> other_arr[0] = 1
>>> other_arr
array([1., 0.])
>>> empty_arr
array([[0., 1., 0.],
       [0., 0., 0.]])
```

maybe on this line, we should assign flux_err_arry to just `np.zeros(img_cube.shape)` to be sure it's empty?

maybe we're not changing `cadence_array` so this isn't an issue?

Comment (falkben) [reply astrocut/cube_cut.py:732]: another think i'm wondering -- should the error be zero? or should it be np.NaN?

Comment (scfleming) [reply astrocut/cube_cut.py:732]: I thought about this too.  I lean towards 0. for two reeasons at the moment:
- the TICA cubes had the dummy values sent to 0. I think, so this maintains that part of it
- adding NaN may result in issues for non-NaN-safe calculations, e.g., produce values of NaN for computation that included the flux uncertainty.  Of course 0. could also introduce problems depending on what the user is doing with it, it's not obious to me which is better if someone isn't aware that these are essentially undefined but use the flux uncertainties in calcuations within their code.

Comment (falkben) [reply astrocut/cube_cut.py:732]: in that case wouldn't it be better to nullify the results than to silently make the calculation with an incorrect value for the error?

Comment (scfleming) [reply astrocut/cube_cut.py:732]: that was my old philosophy: use NaN's as a forced flag to indicate to an unaware user that something is not defined, and you either need to mask or add nan-safe options to your function calls.  these days I wonder how many functions filter out NaN values automatically or not.  Do you have thoughts on this @hmlewis-astro @jaymedina ?

**Comment by hmlewis-astro** [comment]
I agree, keeping the `FLUX_ERR` in the TICA cutouts is fine, and will make things easier for users that want to run the same/similar code on cutouts made from SPOC and TICA cubes.

File: astrocut/cube_cut.py
Comment (scfleming) [reply astrocut/cube_cut.py:732]: +1 on the NaN's.  I just ask we make sure they are filled in such that they are read in correctly with standard astropy fits reading functions, e.g., we don't want to somehow fill them with NaN's and then have them treated as strings when default astropy functions are called on the fits products.  It's probably as easy as making sure the dtype for the FITS table column is correct and you feed them proper np.nan values I guess, although there's a small part of me that worries that having NaN's as not-strings violates a FITS table rule....of course you could then say "we'll just fill that column as strings that say NaN" but then you'll have different dtypes for TICA vs. SPOC in that column...

Comment (jaymedina) [reply astrocut/cube_cut.py:732]: Now that Ben mentions this, I'm also thinking people might assume the 0s are not fillers, especially since in the metadata we explicitly put the unit of the error array as being in electrons ([see here](https://github.com/spacetelescope/astrocut/blob/49dee429c4226de58c74d889915b454e0a17a9ab/astrocut/cube_cut.py#L731)). What we could do is instead fill the array with some ridiculously large number like 9999999 so that users know straight away that this arrays for TICA are filler arrays? Just a thought. Otherwise, I'm leaning towards making them NaN arrays now.

Comment (scfleming) [reply astrocut/cube_cut.py:732]: @jaymedina If we go that route (non-zero but valid number) I would vote for a rediculously large NEGATIVE number (-9999999.0 or something), since negative flux uncertainties would also be treated as suspcious, but in theory you could have crazy large positive values for uncertainties.

Comment (jaymedina) [reply astrocut/cube_cut.py:732]: I'm in favor of that. I'll go ahead and make a commit unless there are other suggestions!

Comment (falkben) [reply astrocut/cube_cut.py:732]: I am not in favor of using any actual number here -- the error/uncertainty is unknown. using actual numbers is likely just going to result in confusion. if we're returning an array to conform with some expectation that the cutout has a certain number of extensions or whatever, but the values are not to be used -- well, this is precisely what np.NaN is for. we might also consider writing something in the header of the fits file about it?

Comment (jaymedina) [reply astrocut/cube_cut.py:732]: I'm not opposed to changing the vals to `np.nan`s. One thing to consider tho, is that `np.zero` arrays are used for all the other columns which we don't have information about, for example `CADENCEN0` which is the column that was referenced in the start of this thread (when making references to `cadence_array`). So if 1 column uses `np.nan` to fill in values we don't know about, and 1 column uses `np.zero`, that will be even more confusing.

What we can do for now is stick with the traditional `np.zero` array (I can reassign `flux_err_array` to `np.zeros(img_cube.shape)` to ensure its a zero array as suggested in the first comment of the thread), and once we come to can agreement on what the fill vals should be, I'll open up a new PR to update the `empty_arr` for all relevant columns.

As for timeline, I'd like to get this PR merged in by end of this week or beginning of next week so I have enough time to update and redeploy the TESSCut API before the end of this sprint.

Comment (hmlewis-astro) [reply astrocut/cube_cut.py:732]: I think this solution works for now! Once you make the changes to reassign `flux_err_array` following Ben's original comment, I think this is ready to merge

Comment (jaymedina) [reply astrocut/cube_cut.py:732]: Sounds good, thanks @hmlewis-astro !

File: astrocut/cube_cut.py
Comment (jaymedina) [inline astrocut/cube_cut.py:723]: There are other calls to `error_arr` besides flux_err_array, so I decided to just remove the assignment of error_array to cadence_array entirely. @falkben

File: astrocut/cube_cut.py
Comment (falkben) [inline astrocut/cube_cut.py:904]: we should probably align on what formatting whitespace formatting we're using on this project to avoid unnecessary changes back and forth...

I use [darker](https://github.com/akaihola/darker) which just applies black selectively to my set of changes (instead of on the entire file).

**Review by falkben — DISMISSED** [review]

**Review by falkben — DISMISSED** [review]

**Comment by jaymedina** [comment]
Sorry, re-requesting because your review got dismissed @falkben

**Review by falkben — APPROVED** [review]

**Review by hmlewis-astro — DISMISSED** [review]

**Review by hmlewis-astro — APPROVED** [review]

**Review by hmlewis-astro — APPROVED** [review]

**Pull request merged**
